### DISCUSSION TITLE
v4 开发计划（v4 Development Plan）

### DISCUSSION THREAD

**Discussion by liuyib** [alibaba/hooks#2405] [General]
## 预期目标

- [ ] 处理完所有  bug  类的  PR  和绝大部分  bug  类  Issue
- [ ] React 18  兼容测试、适配
- [ ] 严格模式下跑通所有测试用例（目前跑严格模式的指令大部分都没有生效，单测文件中 renderHook 需要从 `/packages/hooks/src/utils/tests` 下引入才能测严格模式，而目前大部分是直接从 `@testing-library/react` 引的）
- [ ] 解决存在问题的历史包袱：render  时读/写  ref
  - 相关  issue：[https://github.com/alibaba/hooks/issues/2162](https://github.com/alibaba/hooks/issues/2162)、[https://github.com/alibaba/hooks/issues/780](https://github.com/alibaba/hooks/issues/780)
  - React  文档建议（Do not write or read ref.current during rendering）：[https://react.dev/reference/react/useRef](https://react.dev/reference/react/useRef)
- [ ] 良好的参数校验

## useRequest

- [ ] 所有  open  状态的  Issue/PR：[https://github.com/alibaba/hooks/issues?q=is%3Aopen+in%3Atitle+useRequest+](https://github.com/alibaba/hooks/issues?q=is%3Aopen+in%3Atitle+useRequest+)
- [ ] 轮询重启（需要讨论是否支持）[https://github.com/alibaba/hooks/issues/1827#issuecomment-1794597766](https://github.com/alibaba/hooks/issues/1827#issuecomment-1794597766)
- 整理  useRequest  目前所有待处理的  issue，看有哪些问题，是否需要去修复，有些功能是否可以增强；
- 参考一下  react-query，有一些好的功能可以借鉴过来；

### Scene

- [ ] useVirtualList：overscan  属性设置为  0  时，第一条数据不会渲染  [https://github.com/alibaba/hooks/issues/1538](https://github.com/alibaba/hooks/issues/1538)
	对应 PR（以下 PR 都处理了 #1538 需要再次确认，采用哪个 PR 的方案或综合考虑）：
	- [https://github.com/alibaba/hooks/pull/1808](https://github.com/alibaba/hooks/pull/1808)
	- [https://github.com/alibaba/hooks/pull/2279](https://github.com/alibaba/hooks/pull/2279)
- [ ] useVirtualList：当  itemHeight  为  number  时，list  变化后  offset  计算错误。[https://github.com/alibaba/hooks/pull/2015](https://github.com/alibaba/hooks/pull/2015)
- [ ] useVirtualList:  数据抖动  [https://github.com/alibaba/hooks/pull/1807](https://github.com/alibaba/hooks/pull/1807)
- [x] useCountDown: targetDate resets leftTime [https://github.com/alibaba/hooks/pull/2346](https://github.com/alibaba/hooks/pull/2346)
- [ ] useWebSocket：React 18  严格模式下不可用的问题  [https://github.com/alibaba/hooks/pull/1923](https://github.com/alibaba/hooks/pull/1923)
- [ ] useWebSocket  的  onOpen  问题  [https://github.com/alibaba/hooks/issues/1817](https://github.com/alibaba/hooks/issues/1817)
- [x] useSelections  支持复杂数据类型（或在文档中标注  `T`  不能为对象）：[https://github.com/alibaba/hooks/issues/1105](https://github.com/alibaba/hooks/issues/1105)

### State

- [ ] useResetState：随机初始值  reset  后，没有正常还原的问题  [https://github.com/alibaba/hooks/pull/2562](https://github.com/alibaba/hooks/pull/2562)

### Effect

- [x] useThrottleFn：对  `wait`  参数默认值取值逻辑调整，以支持底层节流防抖的实现细节：[https://github.com/alibaba/hooks/pull/2332](https://github.com/alibaba/hooks/pull/2332)

### Dom

- [ ] useLongPress：滑动过程中触发了长按的回调  [https://github.com/alibaba/hooks/pull/2133](https://github.com/alibaba/hooks/pull/2133)
- [ ] useClickAway：点击目标元素里的动态子元素时，仍触发了回调的问题：[https://github.com/alibaba/hooks/issues/2321](https://github.com/alibaba/hooks/issues/2321)

### Scene

- [ ] useInfiniteScroll：新增  `isInverse`  参数，控制滚动方向  [https://github.com/alibaba/hooks/pull/2315](https://github.com/alibaba/hooks/pull/2315)
- [ ] useInfiniteScroll：强制加载更多的能力  [https://github.com/alibaba/hooks/issues/2180](https://github.com/alibaba/hooks/issues/2180)
- [ ] useDynamicList：新增  `batchRemove`  批量移除的方法   [https://github.com/alibaba/hooks/pull/2340](https://github.com/alibaba/hooks/pull/2340)
- [ ] useDynamicList：允许定制  key  以支持复杂类型的数组  [https://github.com/alibaba/hooks/issues/2280](https://github.com/alibaba/hooks/issues/2280)
- [ ] useVirtualList：添加滚动回调  [https://github.com/alibaba/hooks/pull/2010](https://github.com/alibaba/hooks/pull/2010)
- [ ] useHistoryTravel：支持手动更新   类似于  VueUse  的  [useManualRefHistory](https://vueuse.org/core/useManualRefHistory/)：[https://github.com/alibaba/hooks/pull/2278](https://github.com/alibaba/hooks/pull/2278)
- [ ] useWebSocket:  新增  `binaryType`  参数：[https://github.com/alibaba/hooks/pull/1941](https://github.com/alibaba/hooks/pull/1941)
- [ ] useWebsocket：新增重试回调  [https://github.com/alibaba/hooks/issues/2345](https://github.com/alibaba/hooks/issues/2345)

### State

- [x] useSession/LocalStorageState：存储值变化时，所有  `key`  相同的  `useSession/LocalStorageState`  会同步状态，包括同一浏览器不同  Tab  之间：[https://github.com/alibaba/hooks/pull/2298](https://github.com/alibaba/hooks/pull/2298)
- [ ] 💥 useGetState：同步获取最新状态（之前组件下一次  render  时才能获取最新状态）[https://github.com/alibaba/hooks/pull/2268](https://github.com/alibaba/hooks/pull/2268)
  - [ ] （也可以通过新增  hook，而不引入破坏性）新增  useStateRef hook：[https://github.com/alibaba/hooks/pull/2348](https://github.com/alibaba/hooks/pull/2348)
- [ ] 💥 createUseStorageState：支持批量更新：[https://github.com/alibaba/hooks/pull/2082](https://github.com/alibaba/hooks/pull/2082)，类似需求：
  - [ ] 💥 useCookieState  支持批量更新  [https://github.com/alibaba/hooks/pull/2084](https://github.com/alibaba/hooks/pull/2084)
  - [ ] （破坏性待确认）useRafState  支持批量更新
- [ ] 💥 useUrlState：支持嵌套对象：[https://github.com/alibaba/hooks/issues/2205](https://github.com/alibaba/hooks/issues/2205)
- [ ] 💥 useStorageState/useUrlState/useCookieState  默认值是否应该被持久化？
  - [https://github.com/alibaba/hooks/pull/2064](https://github.com/alibaba/hooks/pull/2064)\- storage

  - [https://github.com/alibaba/hooks/pull/2098](https://github.com/alibaba/hooks/pull/2098)\- cookie

### Dom

- [ ] 💥 useClickAway：空  `target`  数组也允许触发回调：[https://github.com/alibaba/hooks/pull/1953](https://github.com/alibaba/hooks/pull/1953)
- [ ] 💥 useClickAway：支持  Shadow DOM  外的点击事件触发  callback：[https://github.com/alibaba/hooks/pull/2190](https://github.com/alibaba/hooks/pull/2190)
- [ ] useClickAway  的  target  传函数时，支持返回多个  dom target：[https://github.com/alibaba/hooks/issues/1988](https://github.com/alibaba/hooks/issues/1988)
- [ ] 💥 useKeyPress  的兼容性处理，采用  `key`  方案：[https://github.com/alibaba/hooks/pull/2054](https://github.com/alibaba/hooks/pull/2054)
  - 相关  issue：[https://github.com/alibaba/hooks/issues/1467](https://github.com/alibaba/hooks/issues/1467)
  - 成熟库参考：[https://github.com/jamiebuilds/tinykeys](https://github.com/jamiebuilds/tinykeys)
- [ ] useKeyPress:  支持连续按下  N  次按键后再触发回调：[https://github.com/alibaba/hooks/issues/2286](https://github.com/alibaba/hooks/issues/2286)
- [ ] useSize：新增参数以支持  ResizeObserver  的  [options](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver/observe#options)参数：
  - [https://github.com/alibaba/hooks/issues/2227](https://github.com/alibaba/hooks/issues/2227)
  - [https://github.com/alibaba/hooks/issues/2114](https://github.com/alibaba/hooks/issues/2114)

### Advanced

- [ ] useControllableValue：更好的类型提示  [https://github.com/alibaba/hooks/pull/1748](https://github.com/alibaba/hooks/pull/1748)
- [ ] useReactive:  添加安全机制（`options.safe`  参数），组件卸载后不再执行异步回调中的状态变化  [https://github.com/alibaba/hooks/pull/1494](https://github.com/alibaba/hooks/pull/1494)

### Scene

- [ ] useVirtualList：配合  IntersectionObserver  出现抖动：[https://github.com/alibaba/hooks/issues/2324](https://github.com/alibaba/hooks/issues/2324)，有  fork  版解决了这个问题

### Scene

- [ ] useModal：对组件库的弹框组件（Modal、Drawer  等）进行包裹，简化使用、统一代码风格  [https://github.com/alibaba/hooks/pull/2304](https://github.com/alibaba/hooks/pull/2304)，优缺点对比：[https://github.com/alibaba/hooks/issues/2302#issuecomment-1700768460](https://github.com/alibaba/hooks/issues/2302#issuecomment-1700768460)
- [ ] usePostMessage：封装  `postMessage` API，支持跨文档通信  [https://github.com/alibaba/hooks/pull/1977](https://github.com/alibaba/hooks/pull/1977)

### Effect

- [ ] 统一暴露  `createDeepCompareHook`，再把深比较的  hook  都删了：[https://github.com/alibaba/hooks/pull/1972](https://github.com/alibaba/hooks/pull/1972)

### Dom

- [ ] useBounding：封装`getBoundingClientRect` API，支持页面  resize、scroll  时更新  [https://github.com/alibaba/hooks/pull/2063](https://github.com/alibaba/hooks/pull/2063)
- [ ] useSwipe：手势检测  [https://github.com/alibaba/hooks/pull/2337](https://github.com/alibaba/hooks/pull/2337)

## 文档  📝

- [x] 官网生成工具  dumi  升级  v2（目前已在  [chore_dumi_upgrade](https://github.com/alibaba/hooks/tree/chore_dumi_upgrade)分支完成升级，已合并到 v4 分支）
- [ ] useKeyPress  在  Mac  下  Meta  键   和  keyup  的问题  FAQ  文档编写：[https://github.com/alibaba/hooks/pull/2058](https://github.com/alibaba/hooks/pull/2058)
- [ ] useMemoizedFn cannot replace useCallback when functions called during render [https://github.com/alibaba/hooks/discussions/2341](https://github.com/alibaba/hooks/discussions/2341)
- [ ] （🚧 持续更新中）v3->v4 升级文档
- [ ] 英文文档

## 杂项  🛠

- [ ] 项目在  node > 17  版本中无法运行：[https://github.com/alibaba/hooks/issues/2276](https://github.com/alibaba/hooks/issues/2276)
- [x] 解决项目中的  TS Warnings：[https://github.com/alibaba/hooks/pull/2339](https://github.com/alibaba/hooks/pull/2339)
- [x] 升级  query-string  依赖，其底层依赖了有漏洞的  `decode-uri-component`  版本：[https://github.com/alibaba/hooks/pull/2319](https://github.com/alibaba/hooks/pull/2319)
- [x] 升级  prettier from 2.2.0 to 3.0.0 [https://github.com/alibaba/hooks/pull/2248](https://github.com/alibaba/hooks/pull/2248)
- [x]  (❌ 不予处理) 移除打包  es  产物时，多余的  pipe  步骤（貌似不需要过  babel）：[https://github.com/alibaba/hooks/pull/2314](https://github.com/alibaba/hooks/pull/2314)
- [ ] 升级  webpack5，打包产物改为  esnext  以优化体积：[https://github.com/alibaba/hooks/issues/1898](https://github.com/alibaba/hooks/issues/1898)
- [x] 文档站点中，Logo  插入逻辑性能优化：[https://github.com/alibaba/hooks/pull/2307](https://github.com/alibaba/hooks/pull/2307)
- [ ] （🚧 codesandbox 的示例应该是 dumi 自动生成的，待验证升级到 dumi v2 后，这些问题是否解决）文档里的  codesandbox  初始打开后就会报错：
  - 相关 issue：[https://github.com/alibaba/hooks/issues/2226](https://github.com/alibaba/hooks/issues/2226)
  - [https://ahooks.js.org/zh-CN/hooks/use-dynamic-list](https://ahooks.js.org/zh-CN/hooks/use-dynamic-list)\->  第一个  codesandbox  默认不能用
  - [https://ahooks.js.org/zh-CN/hooks/use-selections](https://ahooks.js.org/zh-CN/hooks/use-selections)\->  第一个  codesandbox  默认不能用
- [ ] 创建  Issue Helper？帮助规范提  issue，参考  antd [https://github.com/ant-design/antd-issue-helper](https://github.com/ant-design/antd-issue-helper)

### Scene

- [ ] useInfiniteScroll  异步  bug [https://github.com/alibaba/hooks/issues/2303](https://github.com/alibaba/hooks/issues/2303)
- [ ] useInfiniteScroll  无限滚动加载到第二页以上，执行 reload 数据不会清空而是叠加：[https://github.com/alibaba/hooks/issues/2224](https://github.com/alibaba/hooks/issues/2224)
- [ ] useInfiniteScroll 在切换 tab 场景下执行两次的问题  [https://github.com/alibaba/hooks/issues/1652](https://github.com/alibaba/hooks/issues/1652)
- [ ] usePagination cache cannot use current page to cache pages：[https://github.com/alibaba/hooks/issues/2200](https://github.com/alibaba/hooks/issues/2200)
- [ ] useCountDown  数据错误当本地时间错误时  [https://github.com/alibaba/hooks/issues/1455](https://github.com/alibaba/hooks/issues/1455)
- [ ] next.js  中  useWebSocket options 的  on  事件都不生效  [https://github.com/alibaba/hooks/issues/2309](https://github.com/alibaba/hooks/issues/2309)
- [x] (❌ 不予处理) useWebSocket  优化提议  `socketUrl`  参数支持数组：[https://github.com/alibaba/hooks/issues/2187](https://github.com/alibaba/hooks/issues/2187)

### State

- [ ] useUrlState to use react-router-dom instead of react-router [https://github.com/alibaba/hooks/issues/1725](https://github.com/alibaba/hooks/issues/1725)

### Effect

- [ ] useInterval clear 后，如何重启？ [https://github.com/alibaba/hooks/issues/1829](https://github.com/alibaba/hooks/issues/1829)

### Dom

- [ ] useLongPress  如何可以控制子元素的  onClick  事件  [https://github.com/alibaba/hooks/issues/2325](https://github.com/alibaba/hooks/issues/2325)
- [ ] useLongPress  在内嵌到 ios 的网页中使用时，如果长按结束后弹出一个蒙层，蒙层绑定了 click 事件，在松开手指时会直接触发蒙层的 click 事件[https://github.com/alibaba/hooks/issues/2289](https://github.com/alibaba/hooks/issues/2289)
- [ ] useLongPress  的  moveThreshold  支持对  onClick  生效  [https://github.com/alibaba/hooks/issues/2166](https://github.com/alibaba/hooks/issues/2166)
- [x] (❌ 不予处理) useClickAway  判断不对  [https://github.com/alibaba/hooks/issues/1641](https://github.com/alibaba/hooks/issues/1641)
- [ ] useDrop child trigger avoidance does not work reliably [https://github.com/alibaba/hooks/issues/2036](https://github.com/alibaba/hooks/issues/2036)

### 待分类

- [x] (❌ 不予处理) \[RFC\] usePromiseQueue  将多个并发请求，通过队列形式逐个请求  [https://github.com/alibaba/hooks/issues/2016](https://github.com/alibaba/hooks/issues/2016)
- [ ] \[RFC\] useRightClickMenu / useContextMenu  右键菜单  [https://github.com/alibaba/hooks/issues/1771](https://github.com/alibaba/hooks/issues/1771)
- [ ] \[RFC\] useLoading [https://github.com/alibaba/hooks/issues/1745](https://github.com/alibaba/hooks/issues/1745)

### Dom

- [x] useResponsive：在  Next.js  中导入报错  [https://github.com/alibaba/hooks/issues/2263](https://github.com/alibaba/hooks/issues/2263)（需要  Next.js  来修复）

## 其他

- [ ] 没有记录到本文档的  issue，时间比较久远，从这一页开始往后：[https://github.com/alibaba/hooks/issues?page=4&q=is%3Aissue+is%3Aopen](https://github.com/alibaba/hooks/issues?page=4&q=is%3Aissue+is%3Aopen)

## 竞品参考

- [https://github.com/mantinedev/mantine/tree/master/packages/%40mantine/hooks/src](https://github.com/mantinedev/mantine/tree/master/packages/%40mantine/hooks/src)
- [https://github.com/juliencrn/usehooks-ts/tree/master/packages/usehooks-ts/src](https://github.com/juliencrn/usehooks-ts/tree/master/packages/usehooks-ts/src)
- [https://github.com/uidotdev/usehooks](https://github.com/uidotdev/usehooks/blob/main/index.js)
- [https://github.com/sergeyleschev/react-custom-hooks/tree/main/src/hooks](https://github.com/sergeyleschev/react-custom-hooks/tree/main/src/hooks)
- [~~react-use~~](https://github.com/streamich/react-use) （不再维护，且有很多打破  ref  规则的用法，慎重参考）
- [vue-use](https://github.com/vueuse/vueuse/tree/main/packages/core)

## 总结

- Scene： 15（已分类）+ 8（未分类）
- State： 8（已分类）+ 1（未分类）
- Effect： 2（已分类）+ 1（未分类）
- Dom： 10（已分类）+ 5（未分类）
- Advanced：2（已分类）
- 文档： 3
- 杂项： 9

![image](https://alidocs.oss-cn-zhangjiakou.aliyuncs.com/res/54Lq3Rb1k2V2n7Ed/img/7bf17fe8-0c9e-455a-a418-c34d60f11365.png)

人员分工：

- Scene：1  核心  + 2  社区（任务数占比  45%）
- State+杂项：1  核心  + 1  社区（任务数占比  24%）
- Effect+Dom+Advanced+文档：1  核心  + 2  社区（任务数占比  31%）
- useRequest：3  个核心都参与+ 1  社区

**Comment by liuyib** [comment]
转 @crazylxr 的建议：

> useBounding

有了这个 useSize 是不是可以不需要了

> useSwipe

手势相关的我想把缩放，旋转这些也加进来，加个 usePinch、useRotate

> usePromiseQueue

这个可以改一下，我的想法是加一个 usePromise 的 hook，这里面有更多的基于 promise 扩展的，比如 超时、队列，promisify 等逻辑。

添加 useFPS。

**Comment by KunHuoJ** [comment]
感觉v4 是不是还有点遥遥无期

**Comment by xmsz-stu** [comment]
支持useArray吗？感觉很常用啊

https://github.com/mantinedev/mantine/blob/master/packages/%40mantine/hooks/src/use-list-state/use-list-state.ts
https://github.com/uidotdev/usehooks/blob/90fbbb4cc085e74e50c36a62a5759a40c62bb98e/index.js#L564
https://github.com/sergeyleschev/react-custom-hooks/tree/main/src/hooks/useArray

**Reply by crazylxr** [reply]
https://ahooks.pages.dev/zh-CN/hooks/use-reactive   这个能满足你需求么

**Reply by xmsz-stu** [reply]
> https://ahooks.pages.dev/zh-CN/hooks/use-reactive 这个能满足你需求么

用起来会有点『复杂』，一个改变了赋值习惯，一个是数组会有一些情况很容易忘记比如需要拷贝后才能监听到变化。

**Comment by HaoChenxin** [comment]
V4有计划大约何时发布吗？
